File: 4-parse.py
import json


def move(point):
    point['x'] += offset['x']
    point['y'] += offset['y']
    return point

def conditionalParse(point):
    if str(type(point)) == "<class 'dict'>": return point  # type definition, translation to string
    return json.loads(point)


polyline = [
    {'x': 0, 'y': 0},
    {'x': 10, 'y': 10},
    '{"x": 20,"y": 20}',
    {'x': 30, 'y': 30},
]

print('polyline :', '\n', polyline)
offset = {'x': 10, 'y': -5}
polyline = list(map(conditionalParse, polyline))
path = list(map(move, polyline))
print('path :', '\n', path)

File: 4-parse.js
'use strict';

const move = (offset) => (point) => {
  point.x += offset.x;
  point.y += offset.y;
  return point;
};

const conditionalParse = (item) => {
  // removed the code
  /*
  if (typeof item === 'object') return item;
  return JSON.parse(item);
  */

  // added code
  return (typeof item === 'object' ? item : JSON.parse(item));
};

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const offset = move({ x: 10, y: -5 });
const parsed = polyline.map(conditionalParse);
const path = parsed.map(offset);
console.log({ path });
